### ISSUE TITLE
Jockey teleport bug on Ubuntu versions newer than 18

### ISSUE THREAD

**Issue by LuckyServ** [SirPlease/L4D2-Competitive-Rework#482]
With [Ubuntu 18 hardware and maintenance updates ending in April 2023](https://ubuntu.com/about/release-cycle), we're approaching time to really consider upgrading the OS on which we host servers on.

### Description
A jockeyed survivor sometimes gets teleported in another location, often outside the players space on servers running later versions of Ubuntu.

### Reproduction steps
No reproduction steps so far

### Related fixes
- https://github.com/SirPlease/L4D2-Competitive-Rework/pull/221 (detailed explanation on blood splatter fix by ProdigySim)

### Potential candidates - plugins or extensions that mess with jockey movement
- WaterSlowdown.sp changes some movement flag (m_flLaggedMovementValue) on the jockey when riding a survivor on water, but afaik the bug has happened outside of water
- z_jockey_control_variance removes the ability of survivors to counter movement and I know the dev team had a hard time implementing the control logic so perhaps it can't handle a value of 0 as well as it should

It would be helpful if anyone (perhaps @Target5150) who I heard has servers on Ubuntu 20 can provide patterns / repro steps, as well as other sources of modifications to the jockey movement in this repo.

**Comment by lunatixxx** [comment]
For me it was happening even with the default value of z_jockey_control_variance so it is not that

This plugin fix 90% of the teleport i will say.
https://forums.alliedmods.net/showpost.php?p=2741956&postcount=42

**Comment by Krevik** [comment]
Here it is my version of that plugin cause I am encountering this problem on my servers.
https://github.com/Krevik/Kether.pl-L4D2-Server/blob/kether_2.0/addons/sourcemod/scripting/l4d2_jockey_unteleport.sp

It doesn't have larry's features like map excluding, but I am running the version above since few months on my servers and it worked in 100% cases of jockey teleport.

**Comment by lunatixxx** [comment]
Pretty sure it is not related to ping or alt tab, this happen between two bots too.

**Comment by lunatixxx** [comment]
I just discovered something interesting, i was testing things on my server and i had to go spectator so i did !spec command. Guess what it teleported me out of bounds and triggered the anti jockey teleportation plugin which does not make sense because i was in ready up and no bots. 

Just a guess but the game might think that there is already a jockey on the survivor but when it realises that there is not it brakes and teleport him out of bounds. That will explain why it teleport the survivor when jockeyed, maybe the game think that there is two jockey on the survivor ???

**Comment by lunatixxx** [comment]
I can now safely say it is something related to touch links in the game. I installed this plugin that is supposed to block an exploit on team change: https://github.com/shqke/sp_public/tree/master/remove_touch_links

It litterally triggered 4 teleport in less than one minute even when jockey was dead!

**Comment by lunatixxx** [comment]
These maps are 100% of the time never fixed by jockey_unteleport, only admin can teleport back sadly.

Sacrifice 2 (inside map always be teleported in the middle of the map to the boat)
Cold Stream 1 (out of bounds)

**Comment by SirPlease** [comment]
I appreciate the effort you've put into testing several plugins and methods over the course of time, I do feel like this is something to be fixed by our bigger brained assembly warriors. A plugin to treat the side effects is not optimal in this case. 😓

**Comment by jensewe** [comment]
```cpp
void CTerrorPlayer::OnLeptOnSurvivor(CTerrorPlayer *pVictim)
{
	...
	
	this->m_impactTimer.Start(0.4);
	this->m_impactDuration = 0.4;
	
	...
}

void CTerrorPlayer::UpdateLeap()
{
	...
	
	Vector vecVictimOrigin = pVictim->GetAbsOrigin();
	Vector vecOrigin = this->GetAbsOrigin();
	
	// m_impactTimer, m_flImpactDuration, m_bFinishedImpact
	// No idea what their actual names are, pure guess.
	if ( this->m_impactTimer.HasStarted() && !this->m_impactTimer.IsElasped() )
	{
		float flFactor = 1.0;
		
		if ( this->m_flImpactDuration != 0.0 )
			flFactor = 1.0 - (this->m_impactTimer.GetRemainingTime() / this->m_flImpactDuration);
		
		vecOrigin += (vecVictimOrigin - vecOrigin) * flFactor; // slightly close to victim over time
		this->SetAbsOrigin(vecOrigin);
		
		if ( this->m_bFinishedImpact )
			this->m_bFinishedImpact = false;
	}
	else if ( !this->m_bFinishedImpact )
	{
		this->m_bFinishedImpact = true;
		
		this->SetAbsOrigin(vecVictimOrigin);
		this->SetParent(pVictim, -1);
		
		this->SetMoveType(MOVETYPE_NONE);
	}
	
	...
}
```
Personally no idea what exactly raises the issue, just something I'd like to share.
Convinced myself that in demo playback above codes could lead to a visual issue where a riding jockey dies and gets teleported to somewhere like 0,0,0 (i.e. c2m1_highway, 0,0,0 is mid-air above the final hill if I remember).
And, I kinda believe all the teleport scenes happened within the 0.4s, from what I've witnessed before.

**Comment by Krevik** [comment]
> I appreciate the effort you've put into testing several plugins and methods over the course of time, I do feel like this is something to be fixed by our bigger brained assembly warriors. A plugin to treat the side effects is not optimal in this case. 😓

I totally agree with you, but imo it is better to treat side effects, than letting the bug to destroy a lot of games

**Comment by SirPlease** [comment]
> I totally agree with you, but imo it is better to treat side effects, than letting the bug to destroy a lot of games

Oh a 100%, I'm just less worried about it in the here and now as I don't have this issue as we're still running out stuff on soon to be very oudated Ubuntu. 😁
Sweat droplets are starting to form though, for sure.

**Comment by A1mDev** [comment]
This bug does not depend on linux versions in any way, besides, it is on all servers, since these are shortcomings in the code.

**Comment by ProjectSky** [comment]
> This bug does not depend on linux versions in any way

i don't think so. random teleport never appeared when my server was not upgraded to 20.04.
maybe some newer library triggers this issues?

**Comment by A1mDev** [comment]
Do not forget that 2 bugs with a jockey, they are in different places.

**Comment by SirPlease** [comment]
The bug literally appears the moment you upgrade to anything newer than Ubuntu 18.04 (Didn't check 18.10).
It is **not** on Servers that are running 18.04.

**Comment by SirPlease** [comment]
> Does there is a way to kill an event manually ?

Killing an event doesn't mean it doesn't happen, it just doesn't fire its information forward.
You'll have to block whatever you're trying to block deeper in the code.

> As far i remember (could be wrong) but all teleportations generally happen shortly after round start on first jockey riding, which makes me think this is really related to map change and second round is also considered as a map change by the way. The only thing i don't understand is that even after a server restart it can happen on first map. Maybe if we could cancel jockey ride event on every round start it will help.

It is not tied to round start, Jockeys have been teleporting at the very end of the map just as many times as right away.

**Comment by ProjectSky** [comment]
> > Does there is a way to kill an event manually ?
> 
> Killing an event doesn't mean it doesn't happen, it just doesn't fire its information forward. You'll have to block whatever you're trying to block deeper in the code.
> 
> > As far i remember (could be wrong) but all teleportations generally happen shortly after round start on first jockey riding, which makes me think this is really related to map change and second round is also considered as a map change by the way. The only thing i don't understand is that even after a server restart it can happen on first map. Maybe if we could cancel jockey ride event on every round start it will help.
> 
> It is not tied to round start, Jockeys have been teleporting at the very end of the map just as many times as right away.

correct, it happens to my server many times

**Comment by jensewe** [comment]
> There are some circumstances where jockey ride event does not unhook propely i think. We called a vote to restart the map while a survivor was jockeyed this person was still considered jockeyed for the whole game and caused teleportation bugs on other players.
> 
> Does there is a way to kill an event manually ? https://sm.alliedmods.net/new-api/events/Event/Cancel

To be clear though, event doesn't do anything but provide information at certain time. Guess you'd like to talk about internal `CTerrorPlayer::OnRideEnd()`.
There are actually some other similar cases like charger's queued pummel, when Survivors get charged right before next half begins they are permanently frozen in place on restart.
In this way, it makes sense for jockey teleports with the underlying matter of cleaning state.

**Comment by A1mDev** [comment]
This has nothing to do with the fact that netprop is not reset to 'm_jockeyVictim', otherwise you would hear sound 'JockeyZombie.Ride'. This sound starts when function 'CTerrorPlayer::OnLeptOnSurvivor' is called and ends when function 'CTerrorPlayer::OnRideEnded' is called. This means that the 'CTerrorPlayer::OnRideEnded' function is called. More like a problem with some code in plugins.

**Comment by A1mDev** [comment]
If function 'CTerrorPlayer::OnRideEnded' had not been called you would have run into a lot of problems, you would have noticed it right away. For example, the jockey's ability would not be on cooldown.

**Comment by A1mDev** [comment]
> > The bug literally appears the moment you upgrade to anything newer than Ubuntu 18.04 (Didn't check 18.10). It is **not** on Servers that are running 18.04.
> 
> I wish official servers stopped running on outdated systems, so more people will start complaining about all these bugs. I'll not be surprised that the invisible player one is because of that.

Can you share a demo with this bug?

**Comment by SirPlease** [comment]
Thanks for reporting, fairly certain we can use a much lower unit difference as we're using OnPreThink rather than a previously used method that relied on repeating sourcemod timers, I'll lower it to 400 and see how it goes. 👍

**Comment by ProjectSky** [comment]
maybe more reliable to check for out of bounds using TR_PointOutsideWorld

``` cpp
if (GetVectorDistance(safeVector, preVector) > MAX_SINGLE_FRAME_UNITS || TR_PointOutsideWorld(preVector))
{
	TeleportEntity(client, safeVector, NULL_VECTOR, NULL_VECTOR);
	return;
}
```

**Comment by jensewe** [comment]
I guess it should be known that `sqrt` can be really slow and should be avoided.
So I'm indeed suggesting:
```cpp
#define MAX_SINGLE_FRAME_UNITS_SQUARE 160000.0
if (GetVectorDistance(safeVector, preVector, true) > MAX_SINGLE_FRAME_UNITS_SQUARE) {
    ...
}
```
Hope that I'm not spilting hairs but it's `SDKHooks_PreThink`, as massive as `OnGameFrame` calls, why not?

**Comment by A1mDev** [comment]
Anyway this can't be a bug fix.

**Comment by SirPlease** [comment]
> maybe more reliable to check for out of bounds using TR_PointOutsideWorld

Relying on tracerays doesn't seem like the most cost efficient method.
We're talking about prethink here, which fires constantly, a client moving 400 units within that tiny span is practically impossible and a fairly reliable method.

> Anyway this can't be a bug fix.

Care to elaborate?
It addresses the issue at hand without interfering with the actual game, seems like a fix to me.
